src/cocotb/task.py

```diff
@@ -281,35 +326,27 @@ def _add_done_callback(self, callback: Callable[["Task[ResultType]"], Any]) -> N
             If the task is already done, calling this function will call the callback immediately.
         """
         if self.done():
-            callback(self)
-        self._done_callbacks.append(callback)
+            cocotb._scheduler.instance.schedule(callback, self)
+        else:
+            self._done_callbacks.append(callback)
 
     def __await__(self) -> Generator[Any, Any, ResultType]:
-        # It's tempting to use `return (yield from self._coro)` here,
-        # which bypasses the scheduler. Unfortunately, this means that
-        # we can't keep track of the result or state of the coroutine,
-        # things which we expose in our public API. If you want the
-        # efficiency of bypassing the scheduler, remove the `@coroutine`
-        # decorator from your `async` functions.
-
-        # Hand the coroutine back to the scheduler trampoline.
-        yield self
+        if self._state is TaskState.UNSTARTED:
+            self._schedule_resume()
+        if not self.done():
+            yield self._join
```

Comment:
Avoids redundant checks.
```suggestion
            self._schedule_resume()
            yield self._join
        elif not self.done():
            yield self._join
```